src/backend/joanie/core/models/courses.py

```diff
@@ -343,15 +343,16 @@ def state(self):
         """
         The state of the course carrying information on what to display on a course glimpse.
 
-        The game is to find the highest priority state for this course among its course runs.
+        The game is to find the highest priority state for this course among
+        its course runs and its products.
         """
-        # The default state is for a course that has no course runs
+        # The default state is for a course that has no course runs or products
         best_state = CourseState(CourseState.TO_BE_SCHEDULED)
+        course_runs = self.course_runs.all()
+        products = self.products.all()
 
-        for course_run in self.course_runs.only(
-            "start", "end", "enrollment_start", "enrollment_end"
-        ):
-            state = course_run.state
+        for instance in [*course_runs, *products]:
```

Comment:
This is an opportunity to use itertools chain which will iterate as a generator instead of generating a list in RAM.
It will also save a request on products if the course is open because of its course runs (can be tested!?).

On course runs, you could keep the `only` statement although it does not make much different atm.
On products it's even more usefull to apply an `only` statement with just the `id` field since all you will do is call the `get_equivalent_course_run_dates` method on this model which only does an aggregate?
May be over-ngineered but I'm always concerned by the impact of `all()` queries on the RAM.